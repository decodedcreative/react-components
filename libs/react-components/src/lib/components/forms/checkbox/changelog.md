# Changelog

## 0.1.0

- [MINOR] Initial component creation. Created styling for disabled, checked and error states
