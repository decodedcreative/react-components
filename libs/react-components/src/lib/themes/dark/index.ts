import colors from './colors';
import components from './components';
import general from './general';
import spacing from './spacing';
import typography from './typography';

const darkTheme = {
  name: 'darkTheme',
  colors,
  components,
  general,
  spacing,
  typography,
};

export default darkTheme;
