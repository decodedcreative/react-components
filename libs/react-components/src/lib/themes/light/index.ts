import colors from './colors';
import components from './components';
import general from './general';
import typography from './typography';

const lightTheme = {
  name: 'lightTheme',
  colors,
  components,
  general,
  typography,
};

export default lightTheme;
