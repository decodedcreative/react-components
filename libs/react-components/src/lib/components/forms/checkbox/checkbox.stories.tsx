import { ComponentStory, ComponentMeta } from '@storybook/react';
import { Checkbox } from './checkbox';

export default {
  title: 'Theme/Forms/Checkbox',
  component: Checkbox,
  parameters: {
    backgrounds: {
      default: 'Light',
      values: [
        { name: 'Light', value: '#ffffff' },
        { name: 'Dark', value: '#161b2f' },
      ],
    },
  },
  argTypes: {},
  args: {
    label: 'Would you like to sign up to our mailing list?',
  },
} as ComponentMeta<typeof Checkbox>;

const Template: ComponentStory<typeof Checkbox> = (args) => (
  <Checkbox {...args} />
);

export const Default = Template.bind({});
Default.args = {};

export const Checked = Template.bind({});
Checked.args = {
  checked: true,
};

export const Disabled = Template.bind({});
Disabled.args = {
  disabled: true,
  checked: true,
};

export const Error = Template.bind({});
Error.args = {
  'aria-invalid': true,
  checked: true,
};
