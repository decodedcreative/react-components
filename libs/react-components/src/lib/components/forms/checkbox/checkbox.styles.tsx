import styled from 'styled-components';
import { CheckboxInputProps, CheckboxProps } from './checkbox';
import { Check } from '@styled-icons/bootstrap/';

export const Checkbox = styled('div')`
  position: relative;
  display: flex;
`;

export const CheckboxLabel = styled('label')`
  color: ${(props) => props.theme.colors.color70};
  font-size: ${(props) => props.theme.typography.sizes.default};
  font-family: ${(props) => props.theme.typography.fontFamily};
  line-height: ${(props) => props.theme.typography.lineHeight};
  display: inline-flex;
  align-items: center;
  position: relative;
  z-index: -1;
`;

export const CheckboxBackground = styled('div')`
  background-color: ${(props) => props.theme.colors.color10};
  border: ${(props) => props.theme.general.borderWidth} solid
    ${(props) => props.theme.colors.color50};
  box-sizing: border-box;
  display: flex;
  align-items: center;
  justify-content: center;
  width: 28px;
  height: 28px;
`;

export const CheckboxText = styled('span')`
  margin-left: ${(props) => props.theme.spacing.sm};
`;

export const CheckboxIcon = styled(Check)`
  color: ${(props) => props.theme.colors.interactive};
  width: 24px;
  height: 24px;
`;

export const CheckboxInput = styled('input').attrs<CheckboxInputProps>({
  type: 'checkbox',
})<CheckboxInputProps>`
  position: absolute;
  opacity: 0;
  width: 24px;
  height: 24px;
  border-radius: ${(props) => props.theme.general.borderRadius};

  &:focus ~ ${CheckboxLabel} ${CheckboxBackground} {
    box-shadow: ${(props) => props.theme.general.focusBoxShadow};
    border-color: ${(props) => props.theme.colors.color70};
    outline-width: ${(props) => props.theme.general.focusWidth};
  }

  &:checked::not(&:disabled) ~ ${CheckboxLabel} ${CheckboxBackground} {
    border-color: ${(props) => props.theme.colors.interactive};
  }

  &:disabled ~ ${CheckboxLabel} ${CheckboxBackground} {
    background-color: ${(props) => props.theme.colors.color30};
  }

  &:disabled ~ ${CheckboxLabel} ${CheckboxText} {
    color: ${(props) => props.theme.colors.color50};
  }

  &:disabled ~ ${CheckboxLabel} ${CheckboxIcon} {
    color: ${(props) => props.theme.colors.color50};
  }

  &[aria-invalid='true'] ~ ${CheckboxLabel} ${CheckboxBackground} {
    border-color: ${(props) => props.theme.colors.error};
  }

  &[aria-invalid='true'] ~ ${CheckboxLabel} ${CheckboxIcon} {
    color: ${(props) => props.theme.colors.error};
  }
`;

export const CheckboxIndicator = styled('div')`
  color: ${(props) => props.theme.colors.color10};
  transform: scale(0);
  transition: transform 0.05s ease-in;

  ${CheckboxInput}:checked ~ ${CheckboxLabel} & {
    transform: scale(1);
  }
`;
