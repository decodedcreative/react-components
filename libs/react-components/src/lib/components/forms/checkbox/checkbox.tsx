import { ComponentPropsWithoutRef, forwardRef } from 'react';
import * as S from './checkbox.styles';

export interface CheckboxInputProps extends ComponentPropsWithoutRef<'input'> {}

export interface CheckboxProps extends CheckboxInputProps {
  label: string;
}

export const Checkbox = forwardRef<HTMLInputElement, CheckboxProps>(
  ({ label, ...props }, ref) => {
    const wrapperProps = {
      label,
    };
    const inputProps = {
      ...props,
    };
    return (
      <S.Checkbox {...wrapperProps}>
        <S.CheckboxInput {...inputProps} ref={ref} />
        <S.CheckboxLabel>
          <S.CheckboxBackground>
            <S.CheckboxIndicator>
              <S.CheckboxIcon />
            </S.CheckboxIndicator>
          </S.CheckboxBackground>
          <S.CheckboxText>{label}</S.CheckboxText>
        </S.CheckboxLabel>
      </S.Checkbox>
    );
  }
);
