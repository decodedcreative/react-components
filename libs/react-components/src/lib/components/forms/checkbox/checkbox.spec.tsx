import { render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { lightTheme } from '../../../themes';
import { Checkbox, CheckboxProps } from './checkbox';
import '@testing-library/jest-dom/extend-expect';
import renderer from 'react-test-renderer';
import 'jest-styled-components';

const WithTheme = ({ label, ...props }: CheckboxProps) => (
  <ThemeProvider theme={lightTheme}>
    <Checkbox label={label} {...props} />
  </ThemeProvider>
);

const createComponentJSX = ({ label, ...props }: CheckboxProps) => (
  <WithTheme label={label} {...props} />
);
const createComponentRender = ({ label, ...props }: CheckboxProps) =>
  render(<WithTheme label={label} {...props} />);

describe('Checkbox', () => {
  it('should render successfully', () => {
    const component = createComponentRender({} as CheckboxProps);
    expect(component).toBeTruthy();
  });

  it('should match the default snapshot', () => {
    const defaultCheckbox = createComponentJSX({} as CheckboxProps);
    const defaultCheckboxJSON = renderer.create(defaultCheckbox).toJSON();
    expect(defaultCheckboxJSON).toMatchSnapshot();
  });

  describe('Checkbox - States', () => {
    describe('Checkbox - States - Error', () => {
      it('should have an aria-invalid attribute on its input if in an error state', () => {
        const component = createComponentRender({
          label: 'Checkbox label',
          'aria-invalid': true,
          'data-testid': 'checkbox-input',
        } as CheckboxProps);
        const input = component.getByTestId('checkbox-input');
        expect(input).toHaveAttribute('aria-invalid');
      });
      it('should match the Checkbox With Error snapshot', () => {
        const errorCheckbox = createComponentJSX({
          'aria-invalid': true,
        } as CheckboxProps);
        const errorCheckboxJSON = renderer.create(errorCheckbox).toJSON();
        expect(errorCheckboxJSON).toMatchSnapshot();
      });
    });
    describe('Checkbox - States - Disabled', () => {
      it('should disable its input if a disabled prop is supplied', () => {
        const component = createComponentRender({
          label: 'Checkbox label',
          disabled: true,
          'data-testid': 'checkbox-input',
        } as CheckboxProps);
        const input = component.getByTestId('checkbox-input');
        expect(input).toBeDisabled();
      });
      it('should match the Disabled Checkbox snapshot', () => {
        const disabledCheckbox = createComponentJSX({
          disabled: true,
        } as CheckboxProps);
        const disabledCheckboxJSON = renderer.create(disabledCheckbox).toJSON();
        expect(disabledCheckboxJSON).toMatchSnapshot();
      });
    });
  });

  it('should display its label', () => {
    const component = createComponentRender({
      label: 'Checkbox label 2',
    } as CheckboxProps);
    expect(component.getByText('Checkbox label 2')).toBeTruthy();
  });
});
