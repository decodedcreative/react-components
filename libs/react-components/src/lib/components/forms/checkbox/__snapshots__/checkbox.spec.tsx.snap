// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox Checkbox - States Checkbox - States - Disabled should match the disabled checkbox snapshot 1`] = `
.c9 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c4 {
  color: #272727;
  font-size: 1rem;
  font-family: 'Source Sans Pro',sans-serif;
  line-height: 1.5;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  z-index: -1;
}

.c6 {
  background-color: #ffffff;
  border: 2px solid #8e8e8e;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 28px;
  height: 28px;
}

.c13 {
  margin-left: 0.75rem;
}

.c11 {
  color: #006de3;
  width: 24px;
  height: 24px;
}

.c2 {
  position: absolute;
  opacity: 0;
  width: 24px;
  height: 24px;
  border-radius: 5px;
}

.c2:focus ~ .c3 .c5 {
  box-shadow: 0 0 0 2px #ffffff,0 0 0 4px #006de3;
  border-color: #272727;
  outline-width: 2px;
}

.c2:checked::not(.c1:disabled) ~ .c3 .c5 {
  border-color: #006de3;
}

.c2:disabled ~ .c3 .c5 {
  background-color: #f6f6f6;
}

.c2:disabled ~ .c3 .c12 {
  color: #8e8e8e;
}

.c2:disabled ~ .c3 .c10 {
  color: #8e8e8e;
}

.c2[aria-invalid='true'] ~ .c3 .c5 {
  border-color: #e90014;
}

.c2[aria-invalid='true'] ~ .c3 .c10 {
  color: #e90014;
}

.c8 {
  color: #ffffff;
  -webkit-transform: scale(0);
  -ms-transform: scale(0);
  transform: scale(0);
  -webkit-transition: -webkit-transform 0.05s ease-in;
  -webkit-transition: transform 0.05s ease-in;
  transition: transform 0.05s ease-in;
}

.c1:checked ~ .c3 .c7 {
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
}

<div
  className="c0"
>
  <input
    className="c1 c2"
    disabled={true}
    type="checkbox"
  />
  <label
    className="c3 c4"
  >
    <div
      className="c5 c6"
    >
      <div
        className="c7 c8"
      >
        <svg
          aria-hidden="true"
          className="c9 c10 c11"
          fill="currentColor"
          focusable="false"
          viewBox="0 0 16 16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M10.97 4.97a.75.75 0 0 1 1.07 1.05l-3.99 4.99a.75.75 0 0 1-1.08.02L4.324 8.384a.75.75 0 1 1 1.06-1.06l2.094 2.093 3.473-4.425a.267.267 0 0 1 .02-.022z"
          />
        </svg>
      </div>
    </div>
    <span
      className="c12 c13"
    />
  </label>
</div>
`;

exports[`Checkbox Checkbox - States Checkbox - States - Error should match the checkbox with error snapshot 1`] = `
.c9 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c4 {
  color: #272727;
  font-size: 1rem;
  font-family: 'Source Sans Pro',sans-serif;
  line-height: 1.5;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  z-index: -1;
}

.c6 {
  background-color: #ffffff;
  border: 2px solid #8e8e8e;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 28px;
  height: 28px;
}

.c13 {
  margin-left: 0.75rem;
}

.c11 {
  color: #006de3;
  width: 24px;
  height: 24px;
}

.c2 {
  position: absolute;
  opacity: 0;
  width: 24px;
  height: 24px;
  border-radius: 5px;
}

.c2:focus ~ .c3 .c5 {
  box-shadow: 0 0 0 2px #ffffff,0 0 0 4px #006de3;
  border-color: #272727;
  outline-width: 2px;
}

.c2:checked::not(.c1:disabled) ~ .c3 .c5 {
  border-color: #006de3;
}

.c2:disabled ~ .c3 .c5 {
  background-color: #f6f6f6;
}

.c2:disabled ~ .c3 .c12 {
  color: #8e8e8e;
}

.c2:disabled ~ .c3 .c10 {
  color: #8e8e8e;
}

.c2[aria-invalid='true'] ~ .c3 .c5 {
  border-color: #e90014;
}

.c2[aria-invalid='true'] ~ .c3 .c10 {
  color: #e90014;
}

.c8 {
  color: #ffffff;
  -webkit-transform: scale(0);
  -ms-transform: scale(0);
  transform: scale(0);
  -webkit-transition: -webkit-transform 0.05s ease-in;
  -webkit-transition: transform 0.05s ease-in;
  transition: transform 0.05s ease-in;
}

.c1:checked ~ .c3 .c7 {
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
}

<div
  className="c0"
>
  <input
    aria-invalid={true}
    className="c1 c2"
    type="checkbox"
  />
  <label
    className="c3 c4"
  >
    <div
      className="c5 c6"
    >
      <div
        className="c7 c8"
      >
        <svg
          aria-hidden="true"
          className="c9 c10 c11"
          fill="currentColor"
          focusable="false"
          viewBox="0 0 16 16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M10.97 4.97a.75.75 0 0 1 1.07 1.05l-3.99 4.99a.75.75 0 0 1-1.08.02L4.324 8.384a.75.75 0 1 1 1.06-1.06l2.094 2.093 3.473-4.425a.267.267 0 0 1 .02-.022z"
          />
        </svg>
      </div>
    </div>
    <span
      className="c12 c13"
    />
  </label>
</div>
`;
